lib/common/tests/rules/pcmk__evaluate_op_expression_test.c

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright 2024 the Pacemaker project contributors
+ *
+ * The version control history for this file may have further details.
+ *
+ * This source code is licensed under the GNU General Public License version 2
+ * or later (GPLv2+) WITHOUT ANY WARRANTY.
+ */
+
+#include <crm_internal.h>
+
+#include <stdio.h>
+#include <glib.h>
+
+#include <crm/common/xml.h>
+#include <crm/common/rules_internal.h>
+#include <crm/common/unittest_internal.h>
+#include "crmcommon_private.h"
+
+/*
+ * Shared data
+ */
+
+static pcmk_rule_input_t rule_input = {
+    // These are the only members used to evaluate operation expressions
+    .op_name = PCMK_ACTION_MONITOR,
+    .op_interval_ms = 10000,
+};
+
+/*!
+ * \internal
+ * \brief Run one test, comparing return value
+ *
+ * \param[in] xml_string    Operation expression XML as string
+ * \param[in] reference_rc  Assert that evaluation result equals this
+ */
+static void
+assert_op_expression(const char *xml_string, int reference_rc)
+{
+    xmlNode *xml = pcmk__xml_parse(xml_string);
+
+    assert_int_equal(pcmk__evaluate_op_expression(xml, &rule_input),
+                     reference_rc);
+    free_xml(xml);
+}
+
+
+/*
+ * Invalid arguments
+ */
+
+#define EXPR_NO_MATCH                                       \
+        "<" PCMK_XE_OP_EXPRESSION " " PCMK_XA_ID "='e' "    \
+        PCMK_XA_NAME "='" PCMK_ACTION_START "' "            \
+        PCMK_XA_INTERVAL "='0' />"
+
+static void
+null_invalid(void **state)
+{
+    xmlNode *xml = NULL;
+
+    assert_int_equal(pcmk__evaluate_op_expression(NULL, NULL), EINVAL);
+
+    xml = pcmk__xml_parse(EXPR_NO_MATCH);
+    assert_int_equal(pcmk__evaluate_op_expression(xml, NULL), EINVAL);
+    free_xml(xml);
+
+    assert_op_expression(NULL, EINVAL);
+}
+
+
+/*
+ * Test PCMK_XA_ID
+ */
+
+#define EXPR_ID_MISSING                                 \
+        "<" PCMK_XE_OP_EXPRESSION " "                   \
+        PCMK_XA_NAME "='" PCMK_ACTION_MONITOR "' "      \
+        PCMK_XA_INTERVAL "='10s' />"
+
+static void
+id_missing(void **state)
+{
+    // Currently acceptable
+    assert_op_expression(EXPR_ID_MISSING, pcmk_rc_ok);
+}
+
+
+/*
+ * Test PCMK_XA_NAME
+ */
+
+#define EXPR_NAME_MISSING                                   \
+        "<" PCMK_XE_OP_EXPRESSION " " PCMK_XA_ID "='e' "    \
+        PCMK_XA_INTERVAL "='10s' />"
+
+static void
+name_missing(void **state)
+{
+    assert_op_expression(EXPR_NAME_MISSING, pcmk_rc_unpack_error);
+}
+
+#define EXPR_ALL_MATCH                                      \
+        "<" PCMK_XE_OP_EXPRESSION " " PCMK_XA_ID "='e' "    \
+        PCMK_XA_NAME "='" PCMK_ACTION_MONITOR "' "          \
+        PCMK_XA_INTERVAL "='10s' />"
+
+static void
+reference_name_missing(void **state)
```

Comment:
May want a test for both names missing to ensure that `NULL == NULL` doesn't cause the expression to pass in the future.

Reply:
good idea